lib/ConsumerImpl.cc

```diff
@@ -990,20 +1002,25 @@ void ConsumerImpl::negativeAcknowledge(const MessageId& messageId) {
 
 void ConsumerImpl::disconnectConsumer() {
     LOG_INFO("Broker notification of Closed consumer: " << consumerId_);
-    Lock lock(mutex_);
-    connection_.reset();
-    lock.unlock();
+    resetCnx();
     scheduleReconnection(get_shared_this_ptr());
 }
 
-void ConsumerImpl::closeAsync(ResultCallback callback) {
-    // Keep a reference to ensure object is kept alive
-    ConsumerImplPtr ptr = get_shared_this_ptr();
+void ConsumerImpl::closeAsync(ResultCallback originalCallback) {
+    auto callback = [this, originalCallback](Result result) {
+        shutdown();
```

Comment:
When closing the consumer, we also need to remove the client connection. [cnx->removeConsumer(consumerId_);](https://github.com/apache/pulsar-client-cpp/blob/4a864f20530e230ecee7e4ab09430bd65fb9789c/lib/ConsumerImpl.cc#L1057-L1059)

Also, when invoking `unsubscribe`, don't need to remove the `Connection`, right?

Reply:
I want to determine whether the processing behavior of unsubscribe and close is consistent.

Reply:
> I want to determine whether the processing behavior of unsubscribe and close is consistent.

Good question. I'm also confused when I read Java client. It's very similar, the only difference is, when you failed to unsubscribe a client, you should keep the state as Ready, i.e. no cleanup. https://github.com/apache/pulsar/blob/cbf5cf58196b05a35b31385c6a42098c598c9b33/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L422

Reply:
Yes, I think that after invoking the unsubscribe method, users can still resubscribe. But this is not the case.

Reply:
> When closing the consumer, we also need to remove the client connection

There is an important point that when `HandlerBase::connection_` is modified, we should unregister itself from the connection. Otherwise, the producer or consumer will always be cached in `ClientConnection`. So this PR addes a `beforeConnectionChange` method, which is always called when `setCnx` or `resetCnx` is called.

The `cnx->removeConsumer` is called in `ConsumerImpl::beforeConnectionChange`. Since `shutdown` calls `resetCnx()`, this method is implicitly called in `shutdown`.

Without this design, users have to remove producer or consumer each time the connection changed, like:

```c++
auto cnx = getCnx().lock();
if (cnx) {
    cnx->removeConsumer(consumerId_);
}
connection_ = newConnection;
```

```c++
auto cnx = getCnx().lock();
if (cnx) {
    cnx->removeConsumer(consumerId_);
}
connection_.reset();
```

Reply:
It's the main cause of #55 because the previous code forgot to remove the producer before `connection_.reset()`. When I debugged, I found `cnx` is always empty in the following code block.

https://github.com/apache/pulsar-client-cpp/blob/7f7653b694996f38cf422d8bef24aa1fdfb21c59/lib/ProducerImpl.cc#L720-L723

I didn't look deeper, but it must be caused by the modification of `connection_` field somewhere else.

Reply:
I see. Thanks for your explanation.

Reply:
I think the key point of the unsubscribe operation is to delete the subscription in the broker.

Reply:
I think the key point of the unsubscribe operation is to delete the subscription in the broker.

Reply:
@RobertIndie Yes, this PR just follows what Java client does.

https://github.com/apache/pulsar/blob/cbf5cf58196b05a35b31385c6a42098c598c9b33/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L413-L422